skbio/stats/power.py

```diff
@@ -735,69 +743,49 @@ def paired_subsamples(meta, cat, control_cats, order=None, strict_match=True):
-
-    # Determines the number of samples, and the experimental and control group
-    group_size = np.array([len(cat_groups[o]) for o in order])
-    ctrl_name = order[group_size == group_size.min()][0]
-    order = order[order != ctrl_name]
-
-    # Gets a control group table
-    ctrl_match_groups = meta.groupby(control_cats).groups
-    ctrl_group = meta.loc[cat_groups[ctrl_name]
-                          ].groupby(list(control_cats)).groups
-
-    ids = [np.array([])] * num_groups
-    # Loops through samples in the experimental group to match for controls
-    for check_group, ctrl_ids in viewitems(ctrl_group):
-        # Checks the categories have been defined
-        undefed_check = np.array([_check_strs(p) for p in check_group])
-        if not undefed_check.all() and strict_match:
-            continue
-        # Removes the matched ids from order
-        matched_ids = ctrl_match_groups[check_group]
-        for id_ in ctrl_ids:
-            matched_ids.remove(id_)
-        pos_ids = []
-        num_ids = [len(ctrl_ids)]
-        # Gets the matrix of the matched ids and groups them
-        exp_group = meta.loc[matched_ids].groupby(cat).groups
-        for grp in order:
-            # Checks group to be considered is included in the grouping
-            if grp not in exp_group:
-                break
-            # Gets the id associated with the group
-            pos_ids.append(exp_group[grp])
-            num_ids.append(len(exp_group[grp]))
-        # Determines the minimum number of samples
-        num_draw = np.array(num_ids).min()
-        # Draws samples from possible ids
-        exp_ids = [np.random.choice(ctrl_ids, num_draw, replace=False)]
-        exp_ids.extend([np.random.choice(id_, num_draw, replace=False)
-                        for id_ in pos_ids])
-
-        if len(exp_ids) == num_groups:
-            for idx in range(num_groups):
-                ids[idx] = np.hstack((ids[idx], exp_ids[idx]))
+
+    # Checks the groups in the category
+    min_obs = _get_min_size(meta, cat, control_cats, order, strict_match)
+
+    # Identifies all possible subsamples
+    meta_pairs, index = _identify_sample_groups(meta=meta,
+                                                cat=cat,
+                                                control_cats=control_cats,
+                                                order=order,
+                                                strict_match=strict_match)
+
+    # Draws paired ids
+    ids = _draw_paired_samples(meta_pairs=meta_pairs,
+                               index=index,
+                               num_samps=min_obs)
 
     return ids
 
 
-def _check_strs(x):
-    r"""Returns False if x is a nan and True is x is a string or number"""
+def _get_min_size(meta, cat, control_cats, order, strict_match):
+    """Determines the smallest group represented"""
+    if strict_match:
+        all_cats = copy.deepcopy(control_cats)
+        all_cats.append(cat)
+        meta[all_cats]
```

Comment:
Sorry... I thought you meant the function. Yes, that's left over from debugging.